File: drs/src/test/com/db4o/drs/test/InheritanceTest.java
/* Copyright (C) 2004 - 2006 db4objects Inc. http://www.db4o.com */

package com.db4o.drs.test;

import com.db4o.drs.inside.TestableReplicationProviderInside;

import db4ounit.Assert;


public class InheritanceTest extends DrsTestCase {
	
	private static final String NEW_NAME = "Jeanette";
	private static final String NEW_STUDENTNO = "VANJ";
	private static final int NEW_AGE = 41;
	private static final int OLD_AGE = 37;
	private static final String OLD_STUDENTNO = "VANP";
	private static final String OLD_NAME = "Peter";

	public void Test() {
		Store();
		Replicate(a().provider(), b().provider(), OLD_NAME, OLD_STUDENTNO, OLD_AGE);
		ModifyInB();
		Replicate(b().provider(), a().provider(), NEW_NAME, NEW_STUDENTNO, NEW_AGE);
	}

	private void ModifyInB() {
		Student student = getTheStudent(b().provider());
		ensureOneInstance(b().provider(), Person.class);
		student.setAge(NEW_AGE);
		student.setStudentNo(NEW_STUDENTNO);
		student.setName(NEW_NAME);
		b().provider().update(student);
		b().provider().commit();
		ensureDetails(b().provider(), NEW_NAME, NEW_STUDENTNO, NEW_AGE); 
		ensureDetails(a().provider(), OLD_NAME, OLD_STUDENTNO, OLD_AGE);
	}

	private void Replicate(TestableReplicationProviderInside providerFrom, TestableReplicationProviderInside providerTo, String name, String studentno, int age) {
		replicateAll(providerFrom, providerTo);
		ensureDetails(providerFrom, name, studentno, age);
		ensureDetails(providerTo, name, studentno, age);
	}

	private void Store() {
		Student _student = new Student(OLD_NAME, OLD_AGE);
		_student.setStudentNo(OLD_STUDENTNO);
		a().provider().storeNew(_student);
		a().provider().commit();
		ensureDetails(a().provider(), OLD_NAME, OLD_STUDENTNO, OLD_AGE);
	}
	
	private void ensureDetails(TestableReplicationProviderInside provider, String name, String studentno, int age) {
		ensureOneInstance(provider, Person.class);
		ensureOneInstance(provider, Student.class);
		Person person = getThePerson(provider);
		Assert.areEqual(name, person.getName());
		Assert.areEqual(age, person.getAge());
		Student student = getTheStudent(provider);
		Assert.areEqual(studentno, student.getStudentNo());
		Assert.areEqual(name, student.getName());
		Assert.areEqual(age, student.getAge());
	}

	private Person getThePerson(TestableReplicationProviderInside provider) {
		return (Person) getOneInstance(provider, Person.class);
	}

	private Student getTheStudent(TestableReplicationProviderInside provider) {
		return (Student) getOneInstance(provider, Student.class);
	}
}

File: drs/drs.net/drs.tests/src/com/db4o/drs/test/InheritanceTest.cs
namespace com.db4o.drs.test
{
	public class InheritanceTest : com.db4o.drs.test.DrsTestCase
	{
		private static readonly string NEW_NAME = "Jeanette";

		private static readonly string NEW_STUDENTNO = "VANJ";

		private const int NEW_AGE = 41;

		private const int OLD_AGE = 37;

		private static readonly string OLD_STUDENTNO = "VANP";

		private static readonly string OLD_NAME = "Peter";

		public virtual void Test()
		{
			Store();
			Replicate(A().Provider(), B().Provider(), OLD_NAME, OLD_STUDENTNO, OLD_AGE);
			ModifyInB();
			Replicate(B().Provider(), A().Provider(), NEW_NAME, NEW_STUDENTNO, NEW_AGE);
		}

		private void ModifyInB()
		{
			com.db4o.drs.test.Student student = GetTheStudent(B().Provider());
			EnsureOneInstance(B().Provider(), typeof(com.db4o.drs.test.Person));
			student.SetAge(NEW_AGE);
			student.SetStudentNo(NEW_STUDENTNO);
			student.SetName(NEW_NAME);
			B().Provider().Update(student);
			B().Provider().Commit();
			EnsureDetails(B().Provider(), NEW_NAME, NEW_STUDENTNO, NEW_AGE);
			EnsureDetails(A().Provider(), OLD_NAME, OLD_STUDENTNO, OLD_AGE);
		}

		private void Replicate(com.db4o.drs.inside.TestableReplicationProviderInside providerFrom
			, com.db4o.drs.inside.TestableReplicationProviderInside providerTo, string name, 
			string studentno, int age)
		{
			ReplicateAll(providerFrom, providerTo);
			EnsureDetails(providerFrom, name, studentno, age);
			EnsureDetails(providerTo, name, studentno, age);
		}

		private void Store()
		{
			com.db4o.drs.test.Student _student = new com.db4o.drs.test.Student(OLD_NAME, OLD_AGE
				);
			_student.SetStudentNo(OLD_STUDENTNO);
			A().Provider().StoreNew(_student);
			A().Provider().Commit();
			EnsureDetails(A().Provider(), OLD_NAME, OLD_STUDENTNO, OLD_AGE);
		}

		private void EnsureDetails(com.db4o.drs.inside.TestableReplicationProviderInside 
			provider, string name, string studentno, int age)
		{
			EnsureOneInstance(provider, typeof(com.db4o.drs.test.Person));
			EnsureOneInstance(provider, typeof(com.db4o.drs.test.Student));
			com.db4o.drs.test.Person person = GetThePerson(provider);
			Db4oUnit.Assert.AreEqual(name, person.GetName());
			Db4oUnit.Assert.AreEqual(age, person.GetAge());
			com.db4o.drs.test.Student student = GetTheStudent(provider);
			Db4oUnit.Assert.AreEqual(studentno, student.GetStudentNo());
			Db4oUnit.Assert.AreEqual(name, student.GetName());
			Db4oUnit.Assert.AreEqual(age, student.GetAge());
		}

		private com.db4o.drs.test.Person GetThePerson(com.db4o.drs.inside.TestableReplicationProviderInside
			 provider)
		{
			return (com.db4o.drs.test.Person)GetOneInstance(provider, typeof(com.db4o.drs.test.Person
				));
		}

		private com.db4o.drs.test.Student GetTheStudent(com.db4o.drs.inside.TestableReplicationProviderInside
			 provider)
		{
			return (com.db4o.drs.test.Student)GetOneInstance(provider, typeof(com.db4o.drs.test.Student
				));
		}
	}
}
